tests/Unit/Parallel/Test_AlgorithmLocalSyncAction.cpp

```diff
@@ -0,0 +1,223 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#define CATCH_CONFIG_RUNNER
+
+#include "Framework/TestingFramework.hpp"
+
+#include <cstddef>
+#include <tuple>
+#include <utility>
+#include <vector>
+
+#include "DataStructures/DataBox/DataBox.hpp"
+#include "DataStructures/DataBox/Tag.hpp"
+#include "Parallel/Actions/SetupDataBox.hpp"
+#include "Parallel/Actions/TerminatePhase.hpp"
+#include "Parallel/Algorithms/AlgorithmArray.hpp"
+#include "Parallel/Algorithms/AlgorithmNodegroup.hpp"
+#include "Parallel/GlobalCache.hpp"
+#include "Parallel/InboxInserters.hpp"
+#include "Parallel/InitializationFunctions.hpp"
+#include "Parallel/Main.hpp"
+#include "Parallel/NodeLock.hpp"
+#include "Parallel/ParallelComponentHelpers.hpp"
+#include "Parallel/PhaseDependentActionList.hpp"
+#include "ParallelAlgorithms/Initialization/MutateAssign.hpp"
+#include "Utilities/ErrorHandling/FloatingPointExceptions.hpp"
+#include "Utilities/Gsl.hpp"
+#include "Utilities/System/ParallelInfo.hpp"
+#include "Utilities/TMPL.hpp"
+#include "Utilities/TaggedTuple.hpp"
+
+namespace LocalSyncActionTest {
+template <class Metavariables>
+struct NodegroupComponent;
+
+struct StepNumber : db::SimpleTag {
+  using type = size_t;
+};
+
+struct InitializeNodegroup {
+  template <typename DbTagsList, typename... InboxTags, typename Metavariables,
+            typename ArrayIndex, typename ActionList,
+            typename ParallelComponent>
+  static auto apply(db::DataBox<DbTagsList>& box,
+                    const tuples::TaggedTuple<InboxTags...>& /*inboxes*/,
+                    const Parallel::GlobalCache<Metavariables>& /*cache*/,
+                    const ArrayIndex& /*array_index*/,
+                    const ActionList /*meta*/,
+                    const ParallelComponent* const /*meta*/) noexcept {
+    if constexpr (not tmpl::list_contains_v<DbTagsList, StepNumber>) {
+      return std::make_tuple(
+          db::create_from<db::RemoveTags<>, db::AddSimpleTags<StepNumber>>(
+              std::move(box), 0_st),
+          true);
+    } else {
+      return std::make_tuple(std::move(box), true);
+    }
+  }
+};
+
+/// [sync_action_example]
+struct SyncGetPointerFromNodegroup {
+  using return_type = size_t*;
+  template <typename ParallelComponent, typename DbTagsList>
+  static size_t* apply(db::DataBox<DbTagsList>& box,
+                       const gsl::not_null<Parallel::NodeLock*> node_lock) {
+    if constexpr (tmpl::list_contains_v<DbTagsList, StepNumber>) {
+      size_t* result = nullptr;
+      node_lock->lock();
```

Comment:
Add a comment pointing out that the lock is protecting the DataBox, not the value in the box.